cores/esp8266/core_esp8266_main.cpp

```diff
@@ -400,79 +400,230 @@ extern "C" void __disableWiFiAtBootTime (void)
+extern "C" uint8_t uart_rx_one_char_block();
+
+#if ! FLASH_MAP_SUPPORT
+#include "flash_hal.h"
+#endif
 
 extern "C" void ICACHE_FLASH_ATTR user_pre_init(void)
 {
-    uint32_t rf_cal = 0;
+    const char *flash_map_str = NULL;
+    const char *chip_sz_str = NULL;
+    const char *table_regist_str = NULL;
+    [[maybe_unused]] uint32_t ld_config_chip_size = 0;
+    uint32_t flash_size = 0;
     uint32_t phy_data = 0;
+    uint32_t rf_cal = 0;
     uint32_t system_parameter = 0;
-
-    switch (system_get_flash_size_map())
-    {
-    case FLASH_SIZE_2M:
-        rf_cal = 0x3b000;
-        phy_data = 0x3c000;
-        system_parameter = 0x3d000;
-        break;
-    case FLASH_SIZE_4M_MAP_256_256:
-        rf_cal = 0x7b000;
-        phy_data = 0x7c000;
-        system_parameter = 0x7d000;
-        break;
-    case FLASH_SIZE_8M_MAP_512_512:
-        rf_cal = 0xfb000;
-        phy_data = 0xfc000;
-        system_parameter = 0xfd000;
-        break;
-    case FLASH_SIZE_16M_MAP_512_512:
-    case FLASH_SIZE_16M_MAP_1024_1024:
-        rf_cal = 0x1fb000;
-        phy_data = 0x1fc000;
-        system_parameter = 0x1fd000;
-        break;
-    case FLASH_SIZE_32M_MAP_512_512:
-    case FLASH_SIZE_32M_MAP_1024_1024:
-    case FLASH_SIZE_32M_MAP_2048_2048:
-        rf_cal = 0x3fb000;
-        phy_data = 0x3fc000;
-        system_parameter = 0x3fd000;
-        break;
-    case FLASH_SIZE_64M_MAP_1024_1024:
-        rf_cal = 0x7fb000;
-        phy_data = 0x7fc000;
-        system_parameter = 0x7fd000;
-        break;
-    case FLASH_SIZE_128M_MAP_1024_1024:
-        rf_cal = 0xffb000;
-        phy_data = 0xffc000;
-        system_parameter = 0xffd000;
-        break;
+    [[maybe_unused]] const partition_item_t *_at_partition_table = NULL;
+    size_t _at_partition_table_sz = 0;
+
+    do {
+        #if FLASH_MAP_SUPPORT
+        if (!flashinit()) {
+            flash_map_str = PSTR("flashinit: flash size missing from FLASH_MAP table\n");
+            continue;
+        }
+        #endif
+
+        // For SDKs 3.0.0 and later, place phy_data readonly overlay on top of
+        // the EEPROM address. For older SDKs without a system partition, RF_CAL
+        // and PHY_DATA shared the same flash segment.
+        //
+        // For the Arduino ESP8266 core, the sectors for "EEPROM = size -
+        // 0x5000", "RF_CAL =  size - 0x4000", and "SYSTEM_PARAMETER = size -
+        // 0x3000" are positioned in the last five sectors of flash memory.
+        // PHY_INIT_DATA is special. It is a one time read of 128 bytes of data
+        // that is provided by a spoofed flash read.
+        #if FLASH_MAP_SUPPORT
+        flash_size = __flashdesc[__flashindex].flash_size_kb * 1024u;
+        #else
+        // flashchip->chip_size is updated by the SDK. The size is based on the
+        // value patched into the .bin header by esptool.
+        // system_get_flash_size_map() returns that patched value.
+        flash_size = flashchip->chip_size;
```

Comment:
Rethinking things, should `flash_size` simply be set with `1 << ((spi_flash_get_id() >> 16) & 0xff)` and avoid the possible config selection errors? However, use these config selections in debug builds to verify against the `spi_flash_get_id()` based value.

Reply:
Maybe later down the line, also reducing the number of .ld files
I'd also call debugging as a pretty good reason to keep it as-is